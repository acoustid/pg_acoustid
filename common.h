#ifndef CHROMAPRINT_UTILS_COMMON_H_
#define CHROMAPRINT_UTILS_COMMON_H_

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#ifndef CU_ASSERT
#define CU_ASSERT(x) assert(x)
#endif

#ifndef CU_ALLOC
#define CU_ALLOC(x) malloc(x)
#endif

#ifndef CU_FREE
#define CU_FREE(x) free(x)
#endif

#ifndef CU_REALLOC
#define CU_REALLOC(x, y) realloc(x, y)
#endif

#ifndef CU_LOG_DEBUG
#ifndef NDEBUG
#define CU_LOG_DEBUG(msg, ...) fprintf(stderr, msg, ##__VA_ARGS__); fprintf(stderr, "\n")
#else
#define CU_LOG_DEBUG(msg, ...)
#endif
#endif

#ifndef CU_LOG_INFO
#define CU_LOG_INFO(msg, ...) fprintf(stderr, msg, ##__VA_ARGS__); fprintf(stderr, "\n")
#endif

#ifndef CU_LOG_ERROR
#define CU_LOG_ERROR(msg, ...) fprintf(stderr, msg, ##__VA_ARGS__); fprintf(stderr, "\n")
#endif

#endif /* CHROMAPRINT_UTILS_COMMON_H_ */
