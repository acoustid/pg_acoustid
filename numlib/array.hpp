#pragma once

#include "base.hpp"

#include <iostream>
#include <algorithm>

namespace numlib {

template<typename T>
class Allocator {
 public:
    typedef T value_type;
    typedef T* pointer;
    typedef const T* const_pointer;
    typedef T& reference;
    typedef const T& const_reference;

    Allocator() = default;
    Allocator(const Allocator &other) = default;

    template<typename U>
    Allocator(const Allocator<U> &other) { _ignore_unused(other); }

    pointer alloc(size_t n) {
        return static_cast<pointer>(NUMLIB_ALLOC(n * sizeof(T)));
    }

    pointer realloc(pointer p, size_t n) {
        return static_cast<pointer>(NUMLIB_REALLOC(p, n * sizeof(T)));
    }

    void free(pointer p) {
        NUMLIB_FREE(p);
    }

    template<typename U>
    struct rebind {
        typedef Allocator<U> other;
    };
};

template <typename T, typename Alloc = Allocator<T>>
class Array {
 public:
    typedef T value_type;
    typedef T* iterator;
    typedef const T* const_iterator;
    typedef T* pointer;
    typedef const T* const_pointer;
    typedef T& reference;
    typedef const T& const_reference;
    typedef Alloc allocator_type;

 public:

    Array(const Alloc &alloc = Allocator<T>()) : alloc_(alloc) {}
    Array(size_t n, const Alloc &alloc = Allocator<T>()) : alloc_(alloc) {
        resize(n);
    }
    Array(size_t n, T value, const Alloc &alloc = Allocator<T>()) : alloc_(alloc) {
        resize(n);
        std::fill(begin(), end(), value);
    }
    Array(std::initializer_list<T> init, const Alloc &alloc = Allocator<T>()) : alloc_(alloc) {
        resize(init.size());
        std::copy(init.begin(), init.end(), data_);
    }
    
    ~Array() {
        if (data_) {
            alloc_.free(data_);
        }
    }
    
    Array(const Array& other) : alloc_(other.alloc_), size_(other.size_), capacity_(other.capacity_) {
        data_ = alloc_.alloc(other.size_);
        std::copy(other.data_, other.data_ + other.size_, data_);
    }

    Array& operator=(const Array& other) = delete;
    
    Array(Array&& other) : alloc_(other.alloc_), data_(other.data_), size_(other.size_), capacity_(other.capacity_) {
        other.data_ = nullptr;
        other.size_ = 0;
        other.capacity_ = 0;
    }

    Array& operator=(Array&& other) {
        data_ = other.data_;
        size_ = other.size_;
        capacity_ = other.capacity_;
        other.data_ = nullptr;
        other.size_ = 0;
        other.capacity_ = 0;
        return *this;
    }

    Alloc allocator() const {
        return alloc_;
    }

    reference operator[](size_t i) {
        return data_[i];
    }

    const_reference operator[](size_t i) const {
        return data_[i];
    }

    pointer begin() {
        return data_;
    }

    pointer end() {
        return data_ + size_;
    }

    const_pointer begin() const {
        return data_;
    }

    const_pointer end() const {
        return data_ + size_;
    }

    size_t size() const {
        return size_;
    }

    void resize(size_t n) {
        if (n > capacity_) {
            reserve(n);
        }
        size_ = n;
    }
    
    void reserve(size_t n) {
        if (n <= capacity_) {
            return;
        }
        if (capacity_ == 0) {
            data_ = alloc_.alloc(n);
        } else {
            data_ = alloc_.realloc(data_, n);
        }
        capacity_ = n;
    }

    void push_back(const T& value) {
        Append(value);
    }

    void Append(const T& value) {
        if (size_ == capacity_) {
            auto new_capacity = capacity_ == 0 ? 2 : capacity_ * 2;
            reserve(new_capacity);
        }
        data_[size_++] = value;
    }

 private:
  allocator_type alloc_;
  pointer data_{nullptr};
  size_t size_{0};
  size_t capacity_{0};
};

template <typename T>
class ArrayView {
 public:
    typedef T value_type;
    typedef T* iterator;
    typedef const T* const_iterator;
    typedef T* pointer;
    typedef const T* const_pointer;
    typedef T& reference;
    typedef const T& const_reference;

  ArrayView(T* ptr, size_t size) : ptr_(ptr), size_(size) {}

  T &operator[](size_t i) { return ptr_[i]; }
  const T &operator[](size_t i) const { return ptr_[i]; }

    T* data() { return ptr_; }
    const T* data() const { return ptr_; }

    size_t size() const { return size_; }

    pointer begin() {
        return ptr_;
    }

    pointer end() {
        return ptr_ + size_;
    }

    const_pointer begin() const {
        return ptr_;
    }

    const_pointer end() const {
        return ptr_ + size_;
    }

 private:
    T* ptr_;
    size_t size_;
};

template <typename T>
class ConstArrayView {
 public:
    typedef T value_type;
    typedef T* iterator;
    typedef const T* const_iterator;
    typedef T* pointer;
    typedef const T* const_pointer;
    typedef T& reference;
    typedef const T& const_reference;

  ConstArrayView(const T* ptr, size_t size) : ptr_(ptr), size_(size) {}

  const T &operator[](size_t i) const { return ptr_[i]; }

    const T* data() const { return ptr_; }

    size_t size() const { return size_; }

    const_pointer begin() const {
        return ptr_;
    }

    const_pointer end() const {
        return ptr_ + size_;
    }

 private:
    const T* ptr_;
    size_t size_;
};

template <typename T, typename Deletor>
class OwnedArrayView {
 public:
    OwnedArrayView(T* ptr, size_t size, Deletor d) : ptr_(ptr), size_(size), d_(d) {}

    OwnedArrayView(const OwnedArrayView& o) = delete;
    OwnedArrayView& operator=(const OwnedArrayView& o) = delete;
    void operator=(OwnedArrayView&& o) = delete;
    
    ~OwnedArrayView() {
        d_(ptr_);
    }

    T &operator[](size_t i) { return ptr_[i]; }
    const T &operator[](size_t i) const { return ptr_[i]; }
    
        T* data() { return ptr_; }
        const T* data() const { return ptr_; }
    
        size_t size() const { return size_; }
    
     private:
        T* ptr_;
        size_t size_;
        Deletor d_;
    };

template<typename T>
ArrayView<T> MakeArray(T* ptr, size_t size) {
    return ArrayView<T>(ptr, size);
}

template<typename T>
ConstArrayView<T> MakeArray(const T* ptr, size_t size) {
    return ConstArrayView<T>(ptr, size);
}

template<typename T, typename Deletor>
OwnedArrayView<T, Deletor> MakeArray(T* ptr, size_t size, Deletor d) {
    return OwnedArrayView<T, Deletor>(ptr, size, d);
}

template<typename T, typename Allocator = Allocator<T>>
Array<T> Zeros(size_t n, Allocator allocator = Allocator()) {
    return Array<T>(n, 0, allocator);
}

template<typename T, typename Allocator = Allocator<T>>
Array<T> Ones(size_t n, Allocator allocator = Allocator()) {
    return Array<T>(n, 0, allocator);
}

template<typename T, typename Allocator = Allocator<T>>
Array<T> Range(size_t n, Allocator allocator = Allocator()) {
    auto arr = Array<T>(n, allocator);
    for (size_t i = 0; i < n; ++i) {
        arr[i] = i;
    }
    return arr;
}

template<typename T>
bool operator==(const Array<T>& a, const Array<T>& b) {
    if (a.size() != b.size()) {
        return false;
    }
    for (size_t i = 0; i < a.size(); ++i) {
        if (a[i] != b[i]) {
            return false;
        }
    }
    return true;
}

template<typename T>
std::ostream& operator<<(std::ostream& stream, const Array<T> &a) {
    stream << "{";
    for (size_t i = 0; i < a.size(); ++i) {
        stream << a[i];
        if (i != a.size() - 1) {
            stream << ", ";
        }
    }
    stream << "}";
    return stream;
}

template<typename T>
std::ostream& operator<<(std::ostream& stream, const ArrayView<T> &a) {
    stream << "{";
    for (size_t i = 0; i < a.size(); ++i) {
        stream << a[i];
        if (i != a.size() - 1) {
            stream << ", ";
        }
    }
    stream << "}";
    return stream;
}

} // namespace numlib
