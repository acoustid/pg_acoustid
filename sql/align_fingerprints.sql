CREATE EXTENSION IF NOT EXISTS acoustid;
SET client_min_messages TO DEBUG1;

SELECT aoffset, to_char(ascore, '999D99') AS ascore FROM align_fingerprints(
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint,
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint
);

SELECT aoffset, to_char(ascore, '999D99') AS ascore FROM align_fingerprints(
    '{440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint,
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint
);

SELECT aoffset, to_char(ascore, '999D99') AS ascore FROM align_fingerprints(
    '{440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint,
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint
);
