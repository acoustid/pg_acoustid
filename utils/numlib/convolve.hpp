#pragma once

#include "base.hpp"

namespace numlib {

enum class ConvolveMode {
  FULL,
  SAME,
  VALID
};

template<typename InputType, typename KernelType, typename OutputType>
void Convolve(const InputType &in, const KernelType &kernel, OutputType &out, ConvolveMode mode = ConvolveMode::FULL) {
    size_t in_size = Size(in);
    size_t kernel_size = Size(kernel);
    size_t out_size, in_shift;
    switch (mode) {
        default:
        case ConvolveMode::FULL:
            out_size = in_size + kernel_size - 1;
            in_shift = 0;
            break;
        case ConvolveMode::SAME:
            out_size = in_size;
            in_shift = kernel_size / 2;
            break;
        case ConvolveMode::VALID:
            out_size = in_size - kernel_size + 1;
            in_shift = kernel_size - 1;
            break;
    }
    Resize(out, out_size);
    for (size_t i = 0; i < out_size; ++i) {
        out[i] = 0;
        for (size_t j = 0; j < kernel_size; ++j) {
            size_t k = i + in_shift;
            if (k >= j) {
                k = k - j;
                if (k < in_size) {
                    out[i] += in[k] * kernel[j];
                }
            }
        }
    }
}

template<typename InputType, typename KernelType>
InputType Convolve(const InputType &in, const KernelType &kernel, ConvolveMode mode = ConvolveMode::FULL) {
    InputType out;
    Convolve(in, kernel, out, mode);
    return out;
}

}  // namespace numlib
