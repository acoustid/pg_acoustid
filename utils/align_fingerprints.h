#pragma once

#include <stdint.h>
#include <stddef.h>

#include "common.h"

#ifdef __cplusplus
extern "C" {
#endif

size_t AlignFingerprints(
    const uint32_t *terms1, size_t num_terms1,
    const uint32_t *terms2, size_t num_terms2,
    int *offsets, float *scores, size_t max_offsets,
    int max_distance
);

#ifdef __cplusplus
}
#endif  
