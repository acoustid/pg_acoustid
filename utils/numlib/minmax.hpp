#pragma once

#include "array.hpp"

namespace numlib {

template <typename InputArray, typename Comparator, typename OutputArray>
void ArgRelExtrema(InputArray input, OutputArray &output, Comparator comp, int order = 1) {
    size_t n = Size(input);
    for (size_t i = 0; i < n; i++) {
        auto current_value = input[i];
        int found = 0;
        for (size_t k = 1; k < static_cast<unsigned int>(order) + 1; k++) {
            if (k <= i) {
                if (!comp(current_value, input[i - k])) {
                    found |= 1;
                    break;
                }
            } else {
                found |= 1;
                break;
            }
            if (i + k < n) {
                if (!comp(current_value, input[i + k])) {
                    found |= 2;
                    break;
                }
            } else {
                found |= 2;
                break;
            }
        }
        if (found == 0) {
            output.Append(i);
        }
    }
}

template <typename InputArray, typename OutputArray>
void ArgRelMax(InputArray input, OutputArray &output, int order = 1) {
    ArgRelExtrema(input, output, [](auto a, auto b) { return a > b; }, order);
}

template <typename InputArray, typename OutputArray>
void ArgRelMin(InputArray input, OutputArray &output, int order = 1) {
    ArgRelExtrema(input, output, [](auto a, auto b) { return a < b; }, order);
}

template <typename InputArray, typename T = size_t>
Array<T> ArgRelMax(InputArray input, int order = 1) {
    Array<T> output(input.allocator());
    ArgRelMax(input, output, order);
    return output;
}

template <typename InputArray, typename T = size_t>
Array<T> ArgRelMin(InputArray input, int order = 1) {
    Array<T> output(input.allocator());
    ArgRelMin(input, output, order);
    return output;
}

}
