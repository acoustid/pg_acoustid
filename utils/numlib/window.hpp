#pragma once

#include <cmath>

#include "array.hpp"

namespace numlib {

template<typename T>
void GaussianWindow(T& window, double sigma, bool symetric = true)
{
    auto size = Size(window);
    typename T::value_type center = symetric ? (size - 1) / 2.0 : size / 2.0;
    for (size_t i = 0; i < size; i++)
    {
        window[i] = exp(-0.5 * pow((i - center) / sigma, 2));
    }
}

template<typename T = double>
Array<T> GaussianWindow(size_t size, double sigma, bool symetric = true)
{
    Array<T> window(size);
    GaussianWindow(window, sigma, symetric);
    return window;
}

}
