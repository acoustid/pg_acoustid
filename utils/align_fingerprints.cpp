#include "common.h"
#include "align_fingerprints.h"
#include "numlib/array.hpp"
#include "numlib/convolve.hpp"
#include "numlib/minmax.hpp"
#include "numlib/window.hpp"

namespace nb = numlib;

inline uint32_t PackTermAndOffset(uint16_t term, uint16_t offset) {
    return ((uint32_t)term << 16) | offset;
}

inline bool TermEq(uint32_t a, uint32_t b) {
    return ((a ^ b) & 0xFFFF0000) == 0;
}

inline uint16_t UnpackTerm(uint32_t packed) {
    return (uint16_t)((packed >> 16) & 0xFFFF);
}

inline uint16_t UnpackOffset(uint32_t packed) {
    return (uint16_t)(packed & 0xFFFF);
}

nb::Array<uint32_t> ExpandTerms(const uint32_t *terms, size_t num_terms) {
    nb::Array<uint32_t> expanded_terms;
    expanded_terms.reserve(num_terms * 2);
    for (size_t i = 0; i < num_terms; ++i) {
        expanded_terms.push_back(PackTermAndOffset(terms[i], i));
        expanded_terms.push_back(PackTermAndOffset(terms[i] >> 16, i));
    }
    std::sort(expanded_terms.begin(), expanded_terms.end());
    return expanded_terms;
}

void MergeExpandedTerms(const nb::Array<uint32_t> &terms1, const nb::Array<uint32_t> &terms2, nb::Array<float> &offset_counts, size_t center) {
    size_t i1 = 0, i2 = 0;
    while (i1 < terms1.size() && i2 < terms2.size()) {
        if (TermEq(terms1[i1], terms2[i2])) {
            size_t matching1 = 1;
            size_t matching2 = 1;
            while (i1 + matching1 < terms1.size() && TermEq(terms1[i1 + matching1], terms1[i1])) {
                ++matching1;
            }
            while (i2 + matching2 < terms2.size() && TermEq(terms2[i2 + matching2], terms2[i2])) {
                ++matching2;
            }
            for (size_t j1 = 0; j1 < matching1; ++j1) {
                for (size_t j2 = 0; j2 < matching2; ++j2) {
                    auto offset = center + UnpackOffset(terms1[i1 + j1]) - UnpackOffset(terms2[i2 + j2]);
                    offset_counts[offset] += 1.0 / log(matching1 + matching2);
                }
            }
            i1 += matching1;
            i2 += matching2;
        } else if (terms1[i1] < terms2[i2]) {
            ++i1;
        } else {
            ++i2;
        }
    }
}

size_t AlignFingerprintsImpl(
    const uint32_t *terms1, size_t num_terms1,
    const uint32_t *terms2, size_t num_terms2,
    int *offsets, float *scores, size_t max_offsets,
    int max_distance)
{
    auto expanded_terms1 = ExpandTerms(terms1, num_terms1);
    auto expanded_terms2 = ExpandTerms(terms2, num_terms2);

    // Generate a histogram of the matching offset differences between the two fingerprints.
    auto offset_center = num_terms2;
    auto offset_counts = nb::Zeros<float>(num_terms1 + num_terms2);
    MergeExpandedTerms(expanded_terms1, expanded_terms2, offset_counts, offset_center);

    // Smooth the histogram using a short Gaussian window.
    auto kernel = nb::GaussianWindow(9, 0.6);
    auto smooth_offset_counts = nb::Convolve(offset_counts, kernel, nb::ConvolveMode::SAME);

    // Zero all values below the threshold, which is either 400% the mean or constant 9.0.
    auto threshold = nb::Max(9.0, nb::Mean(smooth_offset_counts) * 4.0);
    for (size_t i = 0; i < smooth_offset_counts.size(); ++i) {
        if (smooth_offset_counts[i] < threshold) {
            smooth_offset_counts[i] = 0.0;
        }
    }

    auto peaks = nb::ArgRelMax(smooth_offset_counts, 5);

    std::sort(peaks.begin(), peaks.end(), [&smooth_offset_counts](size_t i1, size_t i2) {
        return smooth_offset_counts[i1] > smooth_offset_counts[i2];
    });

    size_t num_offsets = 0;
    for (size_t j = 0; j < peaks.size(); ++j) {
        if (num_offsets >= max_offsets) {
            break;
        }
        auto i = peaks[j];
        int offset = i - offset_center;
        if (max_distance >= 0 && abs(offset) > max_distance) {
            continue;
        }
        offsets[num_offsets] = offset;
        scores[num_offsets] = smooth_offset_counts[i];
        num_offsets += 1;
    }
    return num_offsets;
}

size_t AlignFingerprints(
    const uint32_t *terms1, size_t num_terms1,
    const uint32_t *terms2, size_t num_terms2,
    int *offsets, float *scores, size_t max_offsets,
    int max_distance) {
    try {
        return AlignFingerprintsImpl(terms1, num_terms1, terms2, num_terms2, offsets, scores, max_offsets, max_distance);
    } catch (...) {
        return 0;
    }
}
