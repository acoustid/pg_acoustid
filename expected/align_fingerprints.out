CREATE EXTENSION IF NOT EXISTS acoustid;
NOTICE:  extension "acoustid" already exists, skipping
SET client_min_messages TO DEBUG1;
SELECT aoffset FROM align_fingerprints(
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint,
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint
);
DEBUG:  AlignFingerprints returned 1 offsets
 aoffset 
---------
       0
(1 row)

SELECT * FROM align_fingerprints(
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint,
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint
);
DEBUG:  AlignFingerprints returned 1 offsets
 aoffset |  ascore   
---------+-----------
       0 | 23.512262
(1 row)

SELECT * FROM align_fingerprints(
    '{440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint,
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint
);
DEBUG:  AlignFingerprints returned 1 offsets
 aoffset |  ascore   
---------+-----------
      -1 | 20.626871
(1 row)

SELECT * FROM align_fingerprints(
    '{440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint,
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint
);
DEBUG:  AlignFingerprints returned 1 offsets
 aoffset |  ascore  
---------+----------
      -2 | 19.29642
(1 row)

SELECT json_agg(row_to_json(r)) FROM align_fingerprints(
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint,
    '{994277718,440498550,440498550,180443510,180476246,180485494,193052022,193052006,1233231206,1501666678}'::acoustid_raw_fingerprint
) r;
DEBUG:  AlignFingerprints returned 1 offsets
              json_agg              
------------------------------------
 [{"aoffset":0,"ascore":23.512262}]
(1 row)

